File: typescript/src/jobs/manager.ts
import { Job, JobStatus } from "./job.js";

export interface JobManager {
  create(description: string): string;
  start(jobId: string): void;
  complete(jobId: string, result: string): void;
  fail(jobId: string, error: string): void;
  cancel(jobId: string): void;
  get(jobId: string): Job;
  list(status?: JobStatus): Job[];
}

export class InMemoryJobManager implements JobManager {
  private jobs: Map<string, Job> = new Map();
  private counter = 0;
  private lock = new Map<string, true>();

  create(description: string): string {
    this.counter++;
    const jobId = `${Date.now().toString().slice(0, 14)}-${String(this.counter).padStart(4, "0")}`;
    const now = new Date();
    const job: Job = {
      id: jobId,
      status: JobStatus.PENDING,
      description,
      created_at: now,
      updated_at: now,
      result: "",
      error: "",
    };
    this.jobs.set(jobId, job);
    return jobId;
  }

  start(jobId: string): void {
    const job = this.jobs.get(jobId);
    if (!job) throw new Error(`job not found: ${jobId}`);
    job.status = JobStatus.RUNNING;
    job.updated_at = new Date();
  }

  complete(jobId: string, result: string): void {
    const job = this.jobs.get(jobId);
    if (!job) throw new Error(`job not found: ${jobId}`);
    job.status = JobStatus.COMPLETE;
    job.result = result;
    job.updated_at = new Date();
  }

  fail(jobId: string, error: string): void {
    const job = this.jobs.get(jobId);
    if (!job) throw new Error(`job not found: ${jobId}`);
    job.status = JobStatus.FAILED;
    job.error = error;
    job.updated_at = new Date();
  }

  cancel(jobId: string): void {
    const job = this.jobs.get(jobId);
    if (!job) throw new Error(`job not found: ${jobId}`);
    job.status = JobStatus.CANCELED;
    job.updated_at = new Date();
  }

  get(jobId: string): Job {
    const job = this.jobs.get(jobId);
    if (!job) throw new Error(`job not found: ${jobId}`);
    return job;
  }

  list(status?: JobStatus): Job[] {
    if (!status) return Array.from(this.jobs.values());
    return Array.from(this.jobs.values()).filter((j) => j.status === status);
  }
}

File: python/src/pedro_agentware/jobs/manager.py
"""Job manager - Create and track async agent jobs."""

import threading
from datetime import datetime
from typing import Protocol

from .job import Job, JobStatus


class JobManager(Protocol):
    """Protocol for job managers."""

    def create(self, description: str) -> str:
        """Create a new job."""
        ...

    def start(self, job_id: str) -> None:
        """Mark a job as running."""
        ...

    def complete(self, job_id: str, result: str) -> None:
        """Mark a job as complete."""
        ...

    def fail(self, job_id: str, error: str) -> None:
        """Mark a job as failed."""
        ...

    def cancel(self, job_id: str) -> None:
        """Cancel a job."""
        ...

    def get(self, job_id: str) -> Job:
        """Get a job by ID."""
        ...

    def list(self, status: JobStatus | None = None) -> list[Job]:
        """List jobs, optionally filtered by status."""
        ...


class InMemoryJobManager:
    """In-memory implementation of JobManager."""

    def __init__(self) -> None:
        self._jobs: dict[str, Job] = {}
        self._lock = threading.RLock()
        self._counter = 0

    def create(self, description: str) -> str:
        """Create a new job."""
        with self._lock:
            self._counter += 1
            job_id = f"{datetime.now().strftime('%Y%m%d%H%M%S')}-{self._counter:04d}"
            now = datetime.now()
            job = Job(
                id=job_id,
                status=JobStatus.PENDING,
                description=description,
                created_at=now,
                updated_at=now,
            )
            self._jobs[job_id] = job
            return job_id

    def start(self, job_id: str) -> None:
        """Mark a job as running."""
        with self._lock:
            if job_id not in self._jobs:
                raise KeyError(f"job not found: {job_id}")
            job = self._jobs[job_id]
            job.status = JobStatus.RUNNING
            job.updated_at = datetime.now()

    def complete(self, job_id: str, result: str) -> None:
        """Mark a job as complete."""
        with self._lock:
            if job_id not in self._jobs:
                raise KeyError(f"job not found: {job_id}")
            job = self._jobs[job_id]
            job.status = JobStatus.COMPLETE
            job.result = result
            job.updated_at = datetime.now()

    def fail(self, job_id: str, error: str) -> None:
        """Mark a job as failed."""
        with self._lock:
            if job_id not in self._jobs:
                raise KeyError(f"job not found: {job_id}")
            job = self._jobs[job_id]
            job.status = JobStatus.FAILED
            job.error = error
            job.updated_at = datetime.now()

    def cancel(self, job_id: str) -> None:
        """Cancel a job."""
        with self._lock:
            if job_id not in self._jobs:
                raise KeyError(f"job not found: {job_id}")
            job = self._jobs[job_id]
            job.status = JobStatus.CANCELED
            job.updated_at = datetime.now()

    def get(self, job_id: str) -> Job:
        """Get a job by ID."""
        with self._lock:
            if job_id not in self._jobs:
                raise KeyError(f"job not found: {job_id}")
            return self._jobs[job_id]

    def list(self, status: JobStatus | None = None) -> list[Job]:
        """List jobs, optionally filtered by status."""
        with self._lock:
            if status is None:
                return list(self._jobs.values())
            return [j for j in self._jobs.values() if j.status == status]
